System/FilePath/Glob/Base.hs

```diff
@@ -517,7 +519,7 @@ charRange opts zs =
 
 
 optimize :: Pattern -> Pattern
-optimize = liftP (fin . go)
+optimize = liftP (fin . checkUnmatchable . convertExtSeparators . go)
```

Comment:
Can't `convertExtSeparators` be done inside `go`? The way it searches for context that would be directly available in `go` seems somewhat wasteful, on the eyes if not also the CPU. I.e. something like:

```haskell
go (PathSeparator : Literal '.' : _) = [Unmatchable]
go (PathSeparator : ExtSeparator : xs) = PathSeparator : ExtSeparator : go xs
go (ExtSeparator : xs) = Literal '.' : go xs

-- handle the cases where the whole pattern starts with a now-literalized [.]
checkUnmatchable (Literal '.' : _) = [Unmatchable]
checkUnmatchable (LongLiteral _ ('.' : _) : _) = [Unmatchable]
checkUnmatchable ts = if Unmatchable `elem` ts then [Unmatchable] else ts
```

Reply:
Oh of course, I'm not sure why I didn't think of that.

Reply:
With this code, `go` will convert `ExtSeparator` into `Literal '.'` at the very start of a pattern, and then `checkUnmatchable` will convert that into just `[Unmatchable]`, so any pattern starting with a dot, whether explicit or not, becomes `[Unmatchable]`. I think we do need a separate function to deal with `ExtSeparator` at the very start, because the way `go` is used recursively means that we are not able to tell when we are at the beginning of a pattern.

Reply:
With this code, go will convert `ExtSeparator` into `Literal '.'` at the very start of a pattern, and then `checkUnmatchable` will convert that into just `[Unmatchable]`, so any pattern starting with a dot, whether explicit or not, becomes `[Unmatchable]`. I think we do need a separate function to deal with `ExtSeparator` at the very start, because the way `go` is used recursively means that we are not able to tell when we are at the beginning of a pattern.

Reply:
Would it work to make that a separate case at the very top level?

```haskell
optimize (Pattern pat) =
   Pattern . fin $
      case pat of
         ExtSeparator : ts = checkUnmatchable (Literal '.' :) (go ts)
         _ = checkUnmatchable id (go pat)

 where checkUnmatchable f ts = if isUnmatchable ts then [Unmatchable] else f ts
 -- isUnmatchable akin to the checkUnmatchable from my earlier comment but returning Bool
```

Sorry for continuing to shoot from the hip here but `convertExtSeparators` still smells overly complex to me.

Reply:
No, I totally understand; this is just the least complex way I've been able to come up with so far, but I wouldn't go so far as to claim that it's the least complex way full stop. Unfortunately I'm quite busy at the moment so I might not be able to come back to this for a while.

Reply:
That's alright. In that case I'll merge this now, since after all it does fulfill its purpose, and try to simplify it myself when I get a spare moment.

I'll also do a new release next week, whether or not I have time to do something codewise before then. It's already been too long.